cherab/solps/mesh_geometry.py

```diff
@@ -134,6 +134,19 @@ def __init__(self, r, z, vol, neighbix, neighbiy):
         self._triangle_to_grid_map[::2, 1] = xm.flatten()
         self._triangle_to_grid_map[1::2] = self._triangle_to_grid_map[::2]
 
+        # add quadrangle b2 grid
+        self._quadrangles = np.zeros((ng, 4), dtype=np.int32)
+        self._quadrangles[:, 0] = unique_vertices[0:ng]
+        self._quadrangles[:, 1] = unique_vertices[2 * ng: 3 * ng]
+        self._quadrangles[:, 2] = unique_vertices[3 * ng: 4 * ng]
+        self._quadrangles[:, 3] = unique_vertices[ng: 2 * ng]
+        
+        # add mapping from quadrangles to the b2 grid
+        self._quadrangle_to_grid_map = np.zeros((ng, 2), dtype=np.int32)
+        self._quadrangle_to_grid_map[:, 0] = ym.flatten()
+        self._quadrangle_to_grid_map[:, 1] = xm.flatten()
```

Comment:
I think it will be clearer to define the quadrangle grid first and then use it to create the triangle grid:
```python
        ng = self._nx * self._ny  # grid size
        # add quadrangle b2 grid
        self._quadrangles = np.zeros((ng, 4), dtype=np.int32)
        self._quadrangles[:, 0] = unique_vertices[0:ng]
        self._quadrangles[:, 1] = unique_vertices[2 * ng: 3 * ng]
        self._quadrangles[:, 2] = unique_vertices[3 * ng: 4 * ng]
        self._quadrangles[:, 3] = unique_vertices[ng: 2 * ng]

        # add mapping from quadrangles to the b2 grid
        ym, xm = np.meshgrid(np.arange(self._ny, dtype=np.int32), np.arange(self._nx, dtype=np.int32), indexing='ij')
        self._quadrangle_to_grid_map = np.zeros((ng, 2), dtype=np.int32)
        self._quadrangle_to_grid_map[:, 0] = ym.flatten()
        self._quadrangle_to_grid_map[:, 1] = xm.flatten()

        # Number of triangles must be equal to number of rectangle centre points times 2.
        self._num_tris = ng * 2
        self._triangles = np.zeros((self._num_tris, 3), dtype=np.int32)
        self._triangles[0::2, 0] = self._quadrangles[:, 0]
        self._triangles[0::2, 1] = self._quadrangles[:, 1]
        self._triangles[0::2, 2] = self._quadrangles[:, 2]
        # Split the quad cell into two triangular cells.
        self._triangles[1::2, 0] = self._quadrangles[:, 2]
        self._triangles[1::2, 1] = self._quadrangles[:, 3]
        self._triangles[1::2, 2] = self._quadrangles[:, 0]

        # Each triangle cell is mapped to the tuple ID (ix, iy) of its parent mesh cell.
        self._triangle_to_grid_map = np.zeros((self._num_tris, 2), dtype=np.int32)
        self._triangle_to_grid_map[::2, 0] = ym.flatten()
        self._triangle_to_grid_map[::2, 1] = xm.flatten()
        self._triangle_to_grid_map[1::2] = self._triangle_to_grid_map[::2]
```
Because now we can see how the quadrangles are triangulated.

Reply:
Good point, I didn't want to change code which was already there, but this is indeed more instructive. I applied the proposed changes